File: java/MergeKSortedLists.java
import java.util.ArrayList;

public class MergeKSortedLists {
    public ListNode mergeKLists(ArrayList<ListNode> lists) {
        if (lists == null) {
            return null;
        }
        int n = lists.size();
        if (n == 0) {
            return null;
        }
        if (n == 1) {
            return lists.get(0);
        }
        ListNode[] ptrs = new ListNode[n];
        for (int i = 0; i < lists.size(); ++i) {
            ptrs[i] = lists.get(i);
        }
        ListNode head = new ListNode(0);
        ListNode tail = head;
        while (true) {
            boolean flag = true;
            for (int i = 0; i < n; ++i) {
                if (ptrs[i] == null) {
                    flag = false;
                    break;
                }
            }
            if (!flag) {
                break;
            }
            for (int i = 1; i < n; ++i) {
                if (ptrs[i - 1].val != ptrs[i].val) {
                    flag = false;
                    break;
                } else {
                    tail.next = new ListNode(ptrs[i].val);
                    tail = tail.next;
                }
            }
            if (!flag) {
                break;
            }
            for (int i = 0; i < n; ++i) {
                ptrs[i] = ptrs[i].next;
            }
        }
        return head.next;
    }
}

File: cc/merge_k_sorted_lists.cc
#include "base_include.h"
#include "list_node.h"

class MergeKSortedLists {
public:
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        int n = lists.size();
        if (0 == n) {
            return NULL;
        }
        for ( ; 1 < n; n = (n + 1) / 2) {
            for (int i = 0; i < n - 1; i += 2) {
                lists[i / 2] = mergeTwoLists(lists[i], lists[i + 1]);
            }
            if (1 == (n % 2)) {
                lists[n / 2] = lists[n - 1];
            }
        }
        return lists[0];
   }

private:
    ListNode *mergeTwoLists(ListNode *list1, ListNode *list2) {
        ListNode *head = new ListNode(0);
        ListNode *tail = head;
        for ( ; list1 != NULL && list2 != NULL; tail = tail->next) {
            if (list1->val < list2->val) {
                tail->next = list1;
                list1 = list1->next;
            } else {
                tail->next = list2;
                list2 = list2->next;
            }
        }
        for ( ; list1 != NULL; tail = tail->next) {
            tail->next = list1;
            list1 = list1->next;
        }
        for ( ; list2 != NULL; tail = tail->next) {
            tail->next = list2;
            list2 = list2->next;
        }
        ListNode *token = head;
        head = head->next;
        delete token;
        return head;
    }
};

int main(int argc, char *argv[]) {
    return 0;
}
